presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -689,6 +716,27 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
             checkState(rawInputs != null && blockBuilders == null);
             rawInputs.ensureCapacity(rawInputsLength);
 
+            // Perform dedup for distinct aggregates
+            if (delegate instanceof DistinctingGroupedAccumulator) {
```

Comment:
Haven't read the logic yet. A quick design nit:
- instead of doing type check, we can introduce a method for `FinalOnlyGroupedAccumulator` interface called `Page addInputPreprocessing(GroupByIdBlock groupIdsBlock, Page page)` with different implementations for `DistinctingGroupedAccumulator` and `OrderingGroupedAccumulator` (which is a no op)
- `markDistinctHash` is better not visible at `SpillableFinalOnlyGroupedAccumulator` but inside `DistinctingGroupedAccumulator`.
- Would be good to clarify the markDistinctHash already in `DistinctingGroupedAccumulator` and the new one you just introduced. Haven't read the code yet. Are they the same?